pulp_rpm/app/tasks/synchronizing.py

```diff
@@ -26,10 +36,177 @@ def synchronize(remote_pk, repository_pk):
+
+            Args:
+                primary_xml_path(str): a path to a downloaded primary.xml
+                filelists_xml_path(str): a path to a downloaded filelists.xml
+                other_xml_path(str): a path to a downloaded other.xml
+
+            Returns:
+                dict: createrepo_c package objects with the pkgId as a key
+
+            """
+            def pkgcb(pkg):
+                """
+                A callback which is used when a whole package entry in xml is parsed.
+
+                Args:
+                    pkg(preaterepo_c.Package): a parsed metadata for a package
+
+                """
+                packages[pkg.pkgId] = pkg
+
+            def newpkgcb(pkgId, name, arch):
+                """
+                A callback which is used when a new package entry is encountered.
+
+                Only opening <package> element is parsed at that moment.
+                This function has to return a package which parsed data will be added to
+                or None if a package should be skipped.
+
+                pkgId, name and arch of a package can be used to skip further parsing. Available
+                only for filelists.xml and other.xml.
+
+                Args:
+                    pkgId(str): pkgId of a package
+                    name(str): name of a package
+                    arch(str): arch of a package
+
+                Returns:
+                    createrepo_c.Package: a package which parsed data should be added to.
+
+                    If None is returned, further parsing of a package will be skipped.
+
+                """
+                return packages.get(pkgId, None)
+
+            packages = {}
+
+            # TODO: handle parsing errors/warnings, warningcb callback can be used below
+            cr.xml_parse_primary(primary_xml_path, pkgcb=pkgcb, do_files=False)
+            cr.xml_parse_filelists(filelists_xml_path, newpkgcb=newpkgcb)
+            cr.xml_parse_other(other_xml_path, newpkgcb=newpkgcb)
+            return packages
+
+        async def parse_updateinfo(updateinfo_xml_path):
+            """
+            Parse updateinfo.xml to extact update info.
+
+            Args:
+                updateinfo_xml_path: a path to a downloaded updateinfo.xml
+
+            Returns:
+                :obj:`list` of :obj:`createrepo_c.UpdateRecord`: parsed update records
+
+            """
+            uinfo = cr.UpdateInfo()
+
+            # TODO: handle parsing errors/warnings, warningcb callback can be used
+            cr.xml_parse_updateinfo(updateinfo_xml_path, uinfo)
+            return uinfo.updates
+
+        with ProgressBar(message='Downloading and Parsing Metadata') as pb:
+            downloader = self.remote.get_downloader(urljoin(self.remote.url,
+                                                            'repodata/repomd.xml'))
+            # TODO: decide how to distinguish between a mirror list and a normal repo
+            result = await downloader.run()
```

Comment:
`aiohttp.client_exceptions.ServerDisconnectedError: None` DOES NOT happen here  ([my question #4](https://github.com/pulp/pulp_rpm/pull/1160#issuecomment-414135528)).